Documentation/Migrations/Hyperspace 3.x-4.0 Migration Guide.md

```diff
@@ -0,0 +1,111 @@
+
+### Multiple Recovery Strategies
+
+The `BackendService` now supports multiple possible `RecoveryStrategy` objects (previously `RequestRecoveryStrategy`). In the cases where recovery may be applicable, these `RecoveryStrategy` will be queried in order, and the first strategy that is able to attempt a recovery operation will be used. If that recovery attempt then fails, no other strategies will be tried and the request will report a failure result back to the caller.
+
+### Error Handling
+
+There have been several changes made to the core `NetworkService` functionality to allow for more fine grained error handling. These changes include:
+
+- Any time you a custom `Error` type is initialized, the raw `HTTP.Response` received by the server will be provided. This will happen both when a transport failure and decoding failure occurs, and gives access to data like the raw http status and response data.
+- The given `HTTP.Response` includes the `HTTP.Request` that was sent in order to receive that given response.  
+- `DecodingFailureRepresentable` has been changed to provide a `DecodingFailure` which can support more types of errors than just `DecodingError` (for example, a response which should not have been empty but is).
+
+
+## Added Functionality
+
+Along with the core changes described above, there are a ton of new added functionality present in Hyperspace 4.0.
+
+### Swift Package Manager and Carthage Support
+
+A long time coming - but beginning with 4.0, all three of the major dependency managers are supported. Cocoapods, Carthage and SPM - see the README for more details.
+
+### Basic `async` / `await` Support
+
+For users on Xcode 13.2+ and iOS 13+, basic `async` / `await` support has been added in the form of a simple extension on `BackendService`:
+
+```swift
+@available(iOS 13, tvOS 13, watchOS 6, *)
+extension BackendServiceProtocol {
+
+    public func execute<T, U>(request: Request<T, U>) async throws -> T
+
+    public func executeWithResult<T, U>(request: Request<T, U>) async -> Result<T, U>
+}
+```
+
+This is just the beginning of `async` / `await` support in Hyperspace, there will be more to come (and a deeper integration with the `Foundation` `async` APIs) in future versions.
+
+### Request Recovery Additions
+
+In addition to the changes to `RecoveryStrategy` described above, it is also possible to set a `recoveryTransformer` on an individual request - `public var recoveryTransformer: ((TransportFailure) -> TransportSuccess?)`. This property will allow you to attempt to consider any ordinarily failing `Request` as a success. If this handler exists, it is run before any of the `BackendService.recoveryStrategies`. If the handler returns a non-nil `TransportSuccess`, that success will be returned to the caller. If a recovery can not be completed, returning `nil` here will preserve the existing functionality of the `BackendService`.
```

Comment:
Maybe a minor wording change to give a bit more clarity:
```suggestion
In addition to the changes to `RecoveryStrategy` described above, it is also possible to set a `recoveryTransformer` on an individual request - `public var recoveryTransformer: ((TransportFailure) -> TransportSuccess?)`. This property will allow you to attempt to consider any ordinarily failing `Request` as a success. If this handler exists, it is run before any of the `BackendService.recoveryStrategies`. If the handler returns a non-nil `TransportSuccess`, that success will be returned to the caller. If a recovery can not be completed, returning `nil` here will preserve the existing recovery functionality of the `BackendService`.
```

Reply:
Fixed